### PR TITLE
Add kubernetes wrapper docs

### PR DESCRIPTION
**Pull request by mohabusama**
Docs for https://github.com/zalando-zmon/zmon-worker/pull/214

### CODE REVIEW

**Comment by hjacobs** [comment]
Do we really need function aliases like `rs` for replica sets? I would remove these (in the code of course)...

**Comment by mohabusama** [comment]
I thought this would be easier and less error prone (``cm`` vs ``configmaps`` ,`` pv`` vs ``persistentvolumes``), It is also aligned with Kubernetes aliases.

**Comment by hjacobs** [comment]
@mohabusama yes, but reading something like `kubernetes().rs()` in a ZMON check definition looks weird and is hard to understand (for people not knowing Kubernetes details).

**Comment by mohabusama** [comment]
I assumed people using Kubernetes wrapper are already familiar with Kubernetes ;)

I will remove it if you find it confusing :)

**Comment by hjacobs** [comment]
I would remove all short versions/aliases as this is confusing and somebody else (not familiar with Kubernetes) might read the check (e.g. in 24x7 incident case).

**Comment by mohabusama** [comment]
👍

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+.. py:function:: kubernetes(namespace='default')
+
+    ::
+
+    If ``namespace`` is ``None`` then **all** namespaces will be queried. This however will increase the number of calls to Kubernetes API server.
+
+.. note::
+
+    Kubernetes wrapper will authenticate using service account, which assumes the wroker is running in a Kubernetes cluster.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: s/wroker/worker

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+.. _labelSelectors:
+
+Label Selectors
+^^^^^^^^^^^^^^^
+
+Kubernetes API provides a way to filter resources using `labelSelector <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>`_. Kubernetes wrapper provides a friendly syntax for filtering.
+
+The following examples show different usage of Kubernetes wrapper with label filtering:
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: "of the Kubernetes wrapper"

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 1 above)
Comment (lmineiro) [reply docs/user/check-ref/kubernetes_wrapper.rst]: Let's add a short statement clearly saying that all the following functions of the wrapper are scoped to the k8s cluster that the zmon worker belongs to.

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+        Return list of Pods.
+
+        :param name: Pod name.
+        :type name: str
+
+        :param phase: Pod status phase. Valid values are: Pending, Running, Failed, Succeeded or Unknown.
+        :type phase: str
+
+        :param ready: Pod ready status. If None then all pods are returned.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: "Pod Readiness status. If None then all podsare returned"

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of pods. Typical pod has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: nodes(name=None, **kwargs)
+
+        ::
+
+        Return list of nodes. Namespace is ignored.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: "Namespace doesn't apply"

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of nodes. Typical pod has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: services(name=None, **kwargs)
+
+        ::
+
+        Return list of Services.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/services/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+
+Methods of Kubernetes
+^^^^^^^^^^^^^^^^^^^^^
+
+.. py:function:: pods(name=None, phase=None, ready=None, **kwargs)
+
+        ::
+
+        Return list of Pods.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Link to https://kubernetes.io/docs/user-guide/pods/

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 4 above)
Comment (lmineiro) [reply docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/admin/node/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of Endpoints. Typical Endpoint has "metadata", and "subsets".
+        :rtype: list
+
+.. py:function:: ingresses(name=None, **kwargs)
+
+        ::
+
+        Return list of Ingresses.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/ingress/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of Ingresses. Typical Ingress has "metadata", "spec" and "status".
+        :rtype: list
+
+.. py:function:: statefulsets(name=None, replicas=None, **kwargs)
+
+        ::
+
+        Return list of Statefulsets.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/petset/ (or statefulset if they re-fix-f*ck-it-up)

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of Statefulsets. Typical Statefulset has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: replicasets(name=None, replicas=None, **kwargs)
+
+        ::
+
+        Return list of ReplicaSets.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/replicasets/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of ReplicaSets. Typical ReplicaSet has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: deployments(name=None, replicas=None, ready=None, **kwargs)
+
+        ::
+
+        Return list of Deployments.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/deployments/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of Deployments. Typical Deployment has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: configmaps(name=None, **kwargs)
+
+        ::
+
+        Return list of ConfigMaps.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/configmap/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+        :param name: ConfigMap name.
+        :type name: str
+
+        :param **kwargs: ConfigMap :ref:`labelSelectors` filters.
+        :type **kwargs: dict
+
+        :return: List of ConfigMaps. Typical ConfigMap has "metadata", "status" and "spec".
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Couldn't find "spec" in the basic example from k8s docs

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of PersistentVolumeClaims. Typical PersistentVolumeClaim has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: persistentvolumes(name=None, phase=None, **kwargs)
+
+        ::
+
+        Return list of PersistentVolumes.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/persistent-volumes/

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+        :param name: PersistentVolume name.
+        :type name: str
+
+        :param phase: Volume phase.
+        :type phase: str
+
+        :param **kwargs: PersistentVolume :ref:`labelSelectors` filters.
+        :type **kwargs: dict
+
+        :return: List of PersistentVolumes. Typical PersistentVolume has "metadata", "status" and "spec".
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst:234]: Couldn't find "status" in https://kubernetes.io/docs/user-guide/persistent-volumes/ basic example

File: docs/user/check-ref/kubernetes_wrapper.rst [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+        :type **kwargs: dict
+
+        :return: List of ConfigMaps. Typical ConfigMap has "metadata", "status" and "spec".
+        :rtype: list
+
+.. py:function:: persistentvolumeclaims(name=None, phase=None, **kwargs)
+
+        ::
+
+        Return list of PersistentVolumeClaims.
```
Comment (lmineiro) [inline docs/user/check-ref/kubernetes_wrapper.rst]: Add link to https://kubernetes.io/docs/user-guide/persistent-volumes/ (Couldn't find any better link)

**Review by lmineiro — CHANGES_REQUESTED** [review]

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 14 above)
Comment (mohabusama) [reply docs/user/check-ref/kubernetes_wrapper.rst:234]: It is definitely returned by the API

```python
In [1]: pv.obj.keys()
Out[1]: dict_keys(['spec', 'status', 'metadata'])
```

``$ kubectl get pv pvc-1cca9cf3-... -o yaml``

```yaml
...
status:
  phase: Bound
```

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 12 above)
Comment (mohabusama) [reply docs/user/check-ref/kubernetes_wrapper.rst]: Nice catch! only ``metadata`` and ``data`` are returned.

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 1 above)
Comment (mohabusama) [reply docs/user/check-ref/kubernetes_wrapper.rst]: Done

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 2 above)
Comment (mohabusama) [reply docs/user/check-ref/kubernetes_wrapper.rst]: Done

File: docs/user/check-ref/kubernetes_wrapper.rst (hunk 3 above)
Comment (mohabusama) [reply docs/user/check-ref/kubernetes_wrapper.rst]: Done

**Comment by mohabusama** [comment]
@lmineiro all links have been added.

**Comment by lmineiro** [comment]
👍

**Comment by mohabusama** [comment]
👍

**Pull request merged**